### ISSUE TITLE
bug: Multiple medical imaging projects (Brain Tumor Detection, Breast Cancer, Alzheimer's) lack class imbalance handling — models trained on imbalanced datasets report inflated accuracy metrics that mislead learners

### ISSUE THREAD

**Issue by divyanshim27** [abhisheks008/DL-Simplified#1141]
## 🐛 Problem Statement

After reviewing the project structure and notebook patterns across
DL-Simplified's medical imaging projects — specifically `Brain Tumor
Detection/`, `Breast Cancer Detection using DL with WebApp/`,
`Alzheimer's Detection/`, `Blood Disease Detection/`, and
`COVID19-X Ray Image Classification/` — a systematic technical gap
is present across all of them.

Medical imaging datasets are inherently **class-imbalanced**: for example,
a brain tumor dataset might have 80% "No Tumor" and 20% "Tumor" samples.
When a model is trained without addressing this imbalance, a classifier that
**always predicts "No Tumor"** achieves 80% accuracy — which looks
excellent but is medically useless.

### Current Pattern in Affected Notebooks

The notebooks in these projects follow this pattern:

```python
# Typical pattern found in medical imaging notebooks
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(train_generator,
                    epochs=20,
                    validation_data=val_generator)
# Reports: val_accuracy: 0.9247 ← potentially misleading
```

**Missing:** class weight computation, `class_weight` parameter in
`model.fit()`, or oversampling (SMOTE/augmentation on minority class).
**Missing:** evaluation with precision, recall, F1-score, and AUC-ROC
per class — the metrics that actually matter for medical diagnosis.

### Why This Matters for Learners

DL-Simplified is explicitly a **learning resource**. When a beginner
sees 92% accuracy on a brain tumor detector and concludes "this model
is great," they've learned a fundamentally incorrect lesson about how
to evaluate medical AI. This pattern is actively teaching bad practice
at scale across 572 stars and 456 forks.

## ✅ Proposed Solution

For each affected medical imaging project, add a standardized
**Class Imbalance Audit & Fix** section to the notebook:

### Section 1 — Diagnose the imbalance

```python
import numpy as np
from collections import Counter

# Visualize class distribution
label_counts = Counter(train_generator.classes)
print("Class distribution:", label_counts)

# Calculate imbalance ratio
majority = max(label_counts.values())
minority = min(label_counts.values())
print(f"Imbalance ratio: {majority/minority:.1f}:1")

# Plot
import matplotlib.pyplot as plt
plt.bar(label_counts.keys(), label_counts.values())
plt.title("Class Distribution in Training Set")
plt.savefig("../Images/class_distribution.png")
```

### Section 2 — Apply class weights

```python
from sklearn.utils.class_weight import compute_class_weight

class_weights = compute_class_weight(
    class_weight='balanced',
    classes=np.unique(train_generator.classes),
    y=train_generator.classes
)
class_weight_dict = dict(enumerate(class_weights))
print("Class weights:", class_weight_dict)

# Pass to model.fit()
history = model.fit(
    train_generator,
    epochs=20,
    validation_data=val_generator,
    class_weight=class_weight_dict  # ← This line fixes the imbalance issue
)
```

### Section 3 — Use medical-grade evaluation metrics

```python
from sklearn.metrics import (
    classification_report, confusion_matrix,
    roc_auc_score, average_precision_score
)

# Generate predictions on test set
y_pred = model.predict(test_generator)
y_pred_classes = np.argmax(y_pred, axis=1)
y_true = test_generator.classes

# Full report
print(classification_report(
    y_true, y_pred_classes,
    target_names=list(test_generator.class_indices.keys())
))

# AUC-ROC (the metric that actually matters for medical diagnosis)
if y_pred.shape[1] == 2:
    auc = roc_auc_score(y_true, y_pred[:, 1])
    print(f"AUC-ROC: {auc:.4f}")
```

### Files to modify

| Project | Notebook |
|---|---|
| `Brain Tumor Detection/Model/` | Add class weight section |
| `Breast Cancer Detection using DL with WebApp/Model/` | Add class weight section |
| `Alzheimer's Detection/Model/` | Add class weight section |
| `Blood Disease Detection/Model/` | Add class weight section |
| `COVID19-X Ray Image Classification/Model/` | Add class weight section |

Also update the `readme_template.md` to require:
- Class distribution visualization in `Images/`
- F1-score and AUC-ROC reported alongside accuracy in README "Accuracy" section

## 📊 Impact

This is the most consequential technical quality issue in DL-Simplified.
Medical AI that reports 92% accuracy on imbalanced data while ignoring
class weights is incorrect — and teaching this pattern to beginners
produces practitioners who will make the same mistake in real deployments.
Fixing this across the 5 most-starred medical imaging projects sets the
correct standard and can be documented as a teaching moment in each README.

**Suggested labels:** `bug`, `intermediate`, `model-quality`

I would like to work on this. Could you assign it to me?

**Comment by divyanshim27** [comment]
Hey @abhisheks008 ,
I have deeply gone through the project repo with all-proper local running and testing , and now my observations and analysis , I found this as major lagging/enhancement in the project repo .
Based on my issue understanding and my skill-set for solving this problem, I would want to contribute to the issue.
Please assign it to me .

**Comment by Priyanka0205-CSE** [comment]
@abhisheks008 
 I would want to contribute to the issue.
Please assign it to me  under GSSoC 26

**Comment by ParidhiMis** [comment]
i would like to work on this issue. Please /assign it to me under GSSoC

**Comment by MamidiDhana** [comment]
Hi @abhisheks008,
I would like to work on this issue under GSSoC'26.
I have reviewed the problem and understand the importance of handling class imbalance in medical imaging models. I can implement class weight computation, add appropriate evaluation metrics such as Precision, Recall, F1-score, and AUC-ROC, include class distribution visualization, and update the relevant notebooks and documentation while maintaining consistency across the affected projects.
Kindly assign this issue to me. Thank you!

**Comment by karrisanthoshigayatri** [comment]
Hi @abhisheks008,

I'd like to contribute to this issue if it is still available. My approach is to analyze the class distribution in the affected medical imaging datasets, implement class imbalance handling using class weights, evaluate the models with metrics such as Precision, Recall, F1-score, and AUC-ROC, add class distribution visualizations, and update the documentation to promote consistent and reliable model evaluation.

Kindly consider assigning this issue to me if it becomes available. Thank you!
